@xen-orchestra/xapi/vm.mjs

```diff
@@ -399,6 +400,37 @@ class Vm {
     return ref
   }
 
+  async createCloudInitConfig(vmRef, cloudConfig, { networkConfig } = {}) {
+    const vm = await this.getRecord('VM', vmRef)
+
+    // Find the SR of the first VDI.
+    let sr
+    for (const vbdRef of vm.VBDs) {
+      const vbd = await this.getRecord('VBD', vbdRef)
+
+      if (vbd.type !== 'CD' && vbd.VDI !== undefined && Ref.isNotEmpty(vbd.VDI)) {
+        sr = await this.getRecord('SR', await this.getField('VDI', vbd.VDI, 'SR'))
+        break
+      }
+    }
+
+    const {
+      creation: { template: templateUuid },
+    } = extract(vm)
+    const isCoreOsTemplate = (await this.getFieldByUuid('VM', templateUuid, 'name_label')) === 'CoreOS'
+
+    const vmId = vm.uuid
+    const srId = sr.uuid
```

Comment:
Maybe we should put this in the `try` block, so if `sr` is undefined we get the 'createCloudInitConfig failed' warning?

Reply:
Or maybe, before iterating over `VBDs`, a check to see if `vm.VBDs === 0` and throw if so?

Reply:
I think what matters is not really the number of VBDs but the number of VBDs that satisfy `vbd.type !== 'CD' && vbd.VDI !== undefined && Ref.isNotEmpty(vbd.VDI)`
Could we have a VM whose only VBDs don't satisfy the condition?